db/src/tx.rs

```diff
@@ -191,46 +194,146 @@ impl<'conn, 'a> Tx<'conn, 'a> {
     /// The `Term` instances produce share interned TempId and LookupRef handles, and we return the
     /// interned handle sets so that consumers can ensure all handles are used appropriately.
     fn entities_into_terms_with_temp_ids_and_lookup_refs<I>(&self, entities: I) -> Result<(Vec<TermWithTempIdsAndLookupRefs>, intern_set::InternSet<TempId>, intern_set::InternSet<AVPair>)> where I: IntoIterator<Item=Entity> {
-        let mut temp_ids: intern_set::InternSet<TempId> = intern_set::InternSet::new();
-        let mut lookup_refs: intern_set::InternSet<AVPair> = intern_set::InternSet::new();
-
-        let intern_lookup_ref = |lookup_refs: &mut intern_set::InternSet<AVPair>, lookup_ref: entmod::LookupRef| -> Result<LookupRef> {
-            let lr_a: i64 = match lookup_ref.a {
-                entmod::Entid::Entid(ref a) => *a,
-                entmod::Entid::Ident(ref a) => self.schema.require_entid(&a)?,
-            };
-            let lr_attribute: &Attribute = self.schema.require_attribute_for_entid(lr_a)?;
-
-            if lr_attribute.unique.is_none() {
-                bail!(ErrorKind::NotYetImplemented(format!("Cannot resolve (lookup-ref {} {}) with attribute that is not :db/unique", lr_a, lookup_ref.v)))
+        struct InProcess<'a> {
+            schema: &'a Schema,
+            mentat_id_count: i64,
+            temp_ids: intern_set::InternSet<TempId>,
+            lookup_refs: intern_set::InternSet<AVPair>,
+        }
+
+        impl<'a> InProcess<'a> {
+            fn with_schema(schema: &'a Schema) -> InProcess<'a> {
+                InProcess {
+                    schema: schema,
+                    mentat_id_count: 0,
+                    temp_ids: intern_set::InternSet::new(),
+                    lookup_refs: intern_set::InternSet::new(),
+                }
             }
 
-            let lr_typed_value: TypedValue = self.schema.to_typed_value(&lookup_ref.v, &lr_attribute)?;
-            Ok(lookup_refs.intern((lr_a, lr_typed_value)))
-        };
+            fn intern_lookup_ref(&mut self, lookup_ref: &entmod::LookupRef) -> Result<LookupRef> {
+                let lr_a: i64 = match lookup_ref.a {
+                    entmod::Entid::Entid(ref a) => *a,
+                    entmod::Entid::Ident(ref a) => self.schema.require_entid(&a)?,
```

Comment:
Yeah, sorry.  I think it'll be better for you to rebase on this than vice-versa.